### PR TITLE
Geometry_Engine: Split bugs fixed and OutlinesFromLines method added

### LINKED ISSUE #3260: Geometry_Engine: add a method joining unordered lines into closed loops
#### Description:

A feature requested by one of the users, to take a set of unordered, collinear lines and turn them into a set of closed loops that can be extracted, e.g. from the following set of lines:
image

... we would get following 6 outlines returned:
image

...this will also enable reconstruction of meshes from their edges.

### LINKED ISSUE #3270: Infinite loop on Split(Polyline, cuttingLines)
#### Description:

The prophecy has come true. My method managed to not run out of children 😅
https://github.com/BHoM/BHoM_Engine/blob/d13f40bb41726cb0b8ab3b4c809ac6819a8266a5/Geometry_Engine/Compute/Split.cs#L82

image

### LINKED ISSUE #3282: Geometry_Engine: Split(Polyline, List<Line>, double) does not work against chained splitting lines
#### Description:

Due to the way in which `Split(Polyline, List<Line>, double)` method is written, it yields wrong results in cases where splitting curves consist of multiple segments, as in the picture below:
image

In the above case, the splitting curve consists of 3 segments, while the alternative path along the outline is only 2 segments, therefore it is considered 'shorter' by the algorithm, resulting in the splitting curve being simply ignored:
image

After investigation I believe this issue cannot be solved by simply fixing the current implementation, we will need something more complex, based on angles between edges (luckily in my work on https://github.com/BHoM/BHoM_Engine/issues/3260 I may have came up with a faster solution).

Happy to attempt to fix it as I have lots of code ready. Adding @FraserGreenroyd as 2nd assignee because he overviewed the original implementation.

#### Test file(s):
[On SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FIssue%2FBHoM%2FBHoM%5FEngine%2FGeometry%5FEngine%2F%233282%2DSplitBug&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4).

### LINKED ISSUE #3284: Geometry_Engine: BooleanUnion(List<Line>, double, bool) bug in case of multiple overlapping curves
#### Description:

In one of real life scenarios I encountered a bug in `BooleanUnion(List<Line>, double, bool)` method with `bool` input set to `true` - it is related to multiple curves starting and ending at _almost_ same points, which makes `IncludeIntermediatePointsRanges` crash.

Looks like I know how to fix it, so will add it to https://github.com/BHoM/BHoM_Engine/pull/3283.

#### Test file(s):
On [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FBHoM%5FEngine%2FGeometry%5FEngine%2F%233284%2DBooleanUnionBug&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4).

### LINKED ISSUE #3285: Geometry_Engine: Collinearity check is not precise enough
#### Description:

`Query.IsCollinear(List<Point>, double)` is currently a numerical check rather than geometrical - we create a matrix based on point coordinates and check whether it has enough zero rows. Transformation of geometrical tolerance to numerical is virtually impossible, therefore in edge cases points that should be considered collinear within tolerance, are returned as not, which starts becoming an issue in more complex workflows using this method.

This could be fixed by fitting an infinite line between points and then checking if they lie within tolerance from it.

### LINKED ISSUE #3289: Geometry_Engine: Split(Polyline, List<Line>, double) does not work against cutting lines with free ends
#### Description:

As spotted by @vietle-bh , `Split` method throws an exception when the cutting line is freely ended inside the split polygon, as in the picture below:

image

Instead, both cutting lines should be ignored.

#### Steps to reproduce:

See [this comment](https://github.com/BHoM/BHoM_Engine/pull/3283#issuecomment-1953823833)

### PR DESCRIPTION
**Pull request by pawelbaran**
### NOTE: Depends on 
This PR will only work if merged into https://github.com/BHoM/BHoM_Engine/pull/3280 - let the latter get merged first to then get this one rebased.
   
### Issues addressed by this PR

Closes #3260
Closes #3282
Closes #3284
Closes #3285
Closes #3289
Closes #3270 (@vietle-bh **please provide test file**)

### Test files

On [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FBHoM%5FEngine%2FGeometry%5FEngine%2F%233283%2DOutlinesFromLines&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4)

### Additional comments

@FraserGreenroyd could you please have a look at `Split` as the original overviewer? It should be more robust now, edge cases ironed out
@michal-pekacki @adam-sobieski please review `OutlinesFromLines` - this is for your use mainly
@vietle-bh please test this PR against the issue you are having in #3270

I would like to skip UTs in this PR and do them all in one batch under https://github.com/BHoM/BHoM_Engine/issues/3277 once @peterjamesnugent is back, how does that sound @FraserGreenroyd?

### CODE REVIEW

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Ok so I've reviewed the code and run a test and here be some feedback.

Comments on the code have been left at their relevant locations, I don't have any overarching comments regarding the algorithms of themselves at this time so no additional comments there.

On the test however, unfortunately, what's in this PR doesn't work for the original implementation of `Split` 😄 

image

These are 2 models which split happily under the old code.

image

However on this branch I get this _helpful_ error message 😉 (see code comment for my comments on the UX of this error message 😋 ).

So on the basis of the test alone I'm afraid I need to request changes cause it no longer works but did previously 😄 I will upload my model to the linked folder as well for others.

File: Geometry_Engine/Compute/OutlinesFromLines.cs [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+
+        /***************************************************/
+        /****              Private Methods              ****/
+        /***************************************************/
+
+        private static List<Line> CleanUpAndSplitWithEachOther(this List<Line> lines, double distanceTolerance)
+        {
+            Plane fitPlane = lines.SelectMany(x => x.ControlPoints()).ToList().FitPlane();
+            if (fitPlane == null)
+                return null;
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Geometry_Engine/Compute/OutlinesFromLines.cs:86]: The error is not raised on purpose - it is being thrown on the higher level, no need for stacking similar massages on top of each other in my opinion.

File: Geometry_Engine/Compute/OutlinesFromLines.cs [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+        [Description("Combines an unordered set of lines and turns them into closed cells, e.g. mesh edges will be combined into a set of outlines each representing one mesh cell.")]
+        [Input("lines", "Unordered set of lines to turn into closed cells.")]
+        [Input("distanceTolerance", "Tolerance to use for geometrical processing.")]
+        [Output("outlines", "Closed outines created from the input lines.")]
+        public static List<Polyline> OutlinesFromLines(this List<Line> lines, double distanceTolerance = Tolerance.Distance)
+        {
+            lines = lines?.CleanUpAndSplitWithEachOther(distanceTolerance);
+            if (lines == null)
+            {
+                BH.Engine.Base.Compute.RecordError("Outlines could not be created because the input set is not valid.");
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Geometry_Engine/Compute/OutlinesFromLines.cs]: The error is fully gone now after a slight change in code mechanics.

File: Geometry_Engine/Compute/OutlinesFromLines.cs [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+            return lines.ClusterDBSCAN(distanceFunction);
+        }
+
+        /***************************************************/
+
+        private static void RemoveOutliers(this List<Line> lines, double distanceTolerance)
+        {
+            int initialCount = lines.Count;
+
+            Dictionary<Point, int> nodesByValence = new Dictionary<Point, int>();
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Geometry_Engine/Compute/OutlinesFromLines.cs:113]: Comment added 👍

File: Geometry_Engine/Compute/OutlinesFromLines.cs [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+
+                if (!nodesByValence.ContainsKey(l.End))
+                    nodesByValence.Add(l.End, 0);
+
+                nodesByValence[l.End]++;
+            }
+
+            foreach (Point point in nodesByValence.Keys.ToList())
+            {
+                if (nodesByValence.ContainsKey(point))
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Geometry_Engine/Compute/OutlinesFromLines.cs:129]: After the brief chat we had I removed this check without more thinking just to learn that it had been added on purpose 😃 It is needed because `CullValenceOne` method is recursive and may remove some keys from the `nodesByValence` dictionary. So actually it is an extremely important line of code.

File: Geometry_Engine/Compute/OutlinesFromLines.cs [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+
+            // Start at leftmost node
+            Point leftmost = graph.Keys.First();
+            foreach (Point node in graph.Keys)
+            {
+                if (node.X < leftmost.X)
+                    leftmost = node;
+            }
+
+            // Find edge starting at start node with leftmost dir
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Geometry_Engine/Compute/OutlinesFromLines.cs:214]: There is no Y check in this for loop and there had never been 😉 We are looking for the leftmost only point, purely X coordinate based exercise.

File: Geometry_Engine/Compute/OutlinesFromLines.cs [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+                    if (l.Start != currentNode)
+                        dir = dir.Reverse();
+
+                    nodeDirs.Add(dir);
+                }
+
+                Vector newDir = null;
+                Line newEdge = null;
+
+                // To avoid computing heavy signed angle, first check only dirs to the left from the current dir
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Geometry_Engine/Compute/OutlinesFromLines.cs:250]: We are looking for directions to the left from the current direction, i.e. for edges in the left part of the space cut with infinite line defined by current node and current dir.

On an example, let's assume the green rectangle is our space (effectively bounds of points we process), blue is current node and dir. When we position ourselves at the node and look in the direction of dir, we first check for edges in orange area that meet at the node. Only if there is none of them, we check the white area. The reason for it, as in the code comment, is to avoid heave signed angle computations.

image

File: Geometry_Engine/Compute/OutlinesFromLines.cs [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+                    currentEdge = l;
+                    currentDir = dir;
+                }
+            }
+
+            List<Line> outline = new List<Line> { currentEdge };
+            Point currentNode = leftmost;
+
+            // Look for subsequent edges with leftmost dirs compared to the dir of current edge
+            while (true)
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Geometry_Engine/Compute/OutlinesFromLines.cs]: Good point, I rewrote it to leverage `do/while` 👍

File: Geometry_Engine/Compute/OutlinesFromLines.cs [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+                            {
+                                newDir = nodeDirs[j];
+                                newEdge = graph[currentNode][j];
+                            }
+                        }
+                    }
+
+                    if (newEdge == null)
+                    {
+                        BH.Engine.Base.Compute.RecordError("Outlines could not be created due to an internal error.");
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Geometry_Engine/Compute/OutlinesFromLines.cs:401]: It is a fallback, I would say this error should never be hit, but knowing the reality I put a generic message with return `null` to avoid a more serious crash.

**Comment by pawelbaran** [comment]
Thanks for the review @FraserGreenroyd, all comments responded 👍

**Comment by pawelbaran** [comment]
@BHoMBot check required

File: Geometry_Engine/Compute/OutlinesFromLines.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (FraserGreenroyd) [reply Geometry_Engine/Compute/OutlinesFromLines.cs:214]: Line 223 says `currentDir.Y < dir.Y` - should that say `X` then? I know the above for loop on line 210 was handling `X` but this for loop looks like it's handling `Y`?

[the remark this answers is not in the document]
Comment (pawelbaran) [reply Geometry_Engine/Compute/OutlinesFromLines.cs:214]: Hmm, you may be right that I used a mental shortcut here while typing the code comment. Let's make sure we both are on the same page with the process itself:
1. We find leftmost node, i.e. a point lying at the left edge of the bounds
2. From that point, we find the direction looking leftmost in relation to the global Y - since all edges are on the right from the global Y (because we start from the leftmost node), the one with highest Y component will meet the description

So maybe `Find edge starting at leftmost start node with dir closest to global Y` would be more precise?

[the remark this answers is not in the document]
Comment (FraserGreenroyd) [reply Geometry_Engine/Compute/OutlinesFromLines.cs:214]: I think that would make slightly more sense based on the code around the comment to me 👍

**Comment by pawelbaran** [comment]
@BHoMBot check compliance
@BHoMBot check core
@BHoMBot check null-handling

**Comment by pawelbaran** [comment]
@BHoMBot check compliance
@BHoMBot check core
@BHoMBot check null-handling

**Review by FraserGreenroyd — DISMISSED** [review]
My test script is now working, I've reviewed the code and my comments have been addressed satisfactorily, so I'm happy to approve from my side @pawelbaran - happy to take another look if @michal-pekacki finds anything and we need some tweaks but until that moment, here is my approval for the `Split` element for the ceiling tile workflow 😄

**Comment by pawelbaran** [comment]
Thanks @vietle-bh for the review, well spotted! I have raised #3289 to capture the issue, then fixed it in this PR, also included that case in test files linked to this PR.

Finally, I made use of your polygon to stress test the code, with satisfying results 😃 
image

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
One more edge case fix pushed + test added to the test files, the code should exactly as it did just more robust, all tests passing 👍

**Comment by vietle-bh** [comment]
> Thanks @vietle-bh for the review, well spotted! I have raised #3289 to capture the issue, then fixed it in this PR, also included that case in test files linked to this PR.
> 
> Finally, I made use of your polygon to stress test the code, with satisfying results 😃 

I use the existing Split method intensively in a recursive method for a custom tool. The new Split method in this PR hasn't failed for the last several days, and seems to have resolved https://github.com/BHoM/BHoM_Engine/issues/3270 👍 

image

**Comment by vietle-bh** [comment]
Another great thing I got from this PR is a way to break down self-intersecting polylines. Thanks @pawelbaran 😍 
Out of interest, what did everyone previously use for breaking down self-intersecting polylines?

image

**Comment by vietle-bh** [comment]
> Another great thing I got from this PR is a way to break down self-intersecting polylines. Thanks @pawelbaran 😍 Out of interest, what did everyone previously use for breaking down self-intersecting polylines?
> 
Guess what I'm really asking is which way is most performant 😋

**Comment by vietle-bh** [comment]
Great teamwork! By the way, the Split method returns that generic "internal error" message if I feed it the self-intersecting line. It would be great to have it check for self-intersection and give a more specific message.

image

**Comment by pawelbaran** [comment]
> Great teamwork! By the way, the Split method returns that generic "internal error" message if I feed it the self-intersecting line. It would be great to have it check for self-intersection and give a more specific message.
> 
> image

What do you mean by self-intersecting line? The outline?

**Comment by pawelbaran** [comment]
I added more preliminary checks to the `Split` method, also included these cases in the test file @vietle-bh - should be all good now.

@FraserGreenroyd FYI I added `IsNull` overloads in BHoM_Engine and cleaned up the code a bit over there, could be worth you having a quick look 👍

**Review by michal-pekacki — DISMISSED** [review]
Tested `OutlinesFromLines` and it works as it should, happy to approve 👍

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Review by michal-pekacki — APPROVED** [review]
Happy to re-approve 👍

**Comment by pawelbaran** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance
@BHoMBot check unit-tests

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: unit-tests

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge
